File: lib/proxyconf/downstream_auth.lua
local md5 = require("md5")

-- Config gets filled by elixir
Config = {}

local function validate_hash(value, hashes)
  if type(value) == "string" then
    local hash = md5.sumhexa(value)
    print("hash", hash)
    local client = hashes[hash]
    if client ~= nil and client.allow then
      return client
    end
    return nil
  end
end

local function get_query_parameter(query_string, parameter_name)
  for key, val in query_string:gmatch("([^&=?]-)=([^&=?]+)") do
    key = key:gsub("=+.*$", "")
    key = key:gsub("%s", "_") -- remove spaces in parameter name
    if key == parameter_name then
      val = val:gsub("^=+", "")
      return val
    end
  end
  return nil
end

local function to_key(api_id, auth_type, auth_field_name)
  return api_id .. ":" .. auth_type .. ":" .. auth_field_name
end

-- Note: Cache the result of the schema compilation as this is quite expensive
function envoy_on_request(request_handle)
  local metadata = request_handle:metadata()
  local api_id = metadata:get("api_id")
  local auth_type = metadata:get("auth_type")
  local auth_field_name = metadata:get("auth_field_name")

  if api_id == nil or auth_type == nil or auth_field_name == nil then
    request_handle:logInfo(string.format("Downstream authentication disabled for API %s", "disables-api"))
    request_handle:streamInfo():dynamicMetadata():set("proxyconf.downstream_auth", "status", "noauth")
    return
  end

  local hashes = Config[to_key(api_id, auth_type, auth_field_name)]
  if hashes then
    local client = nil
    if auth_type == "header" then
      local header = request_handle:headers():get(auth_field_name)
      client = validate_hash(header, hashes)
    elseif auth_type == "query" then
      local path = request_handle:headers():get(":path")
      local query_param = get_query_parameter(path, auth_field_name)
      client = validate_hash(query_param, hashes)
    end
    if client ~= nil then
      request_handle:streamInfo():dynamicMetadata():set("proxyconf.downstream_auth", "client_id", client.id)
      request_handle:streamInfo():dynamicMetadata():set("proxyconf.downstream_auth", "status", "success")
      request_handle:logInfo(
        string.format(
          "Downstream authentication (%s:%s) success for API %s and client %s",
          auth_type,
          auth_field_name,
          api_id,
          client.id
        )
      )
      -- auth success
      return
    else
      -- auth failed
      request_handle:streamInfo():dynamicMetadata():set("proxyconf.downstream_auth", "status", "failed")
      request_handle:respond({ [":status"] = "403" }, "Forbidden")
      request_handle:logInfo(
        string.format(
          "Downstream authentication (%s:%s) failed for API %s and client %s",
          auth_type,
          auth_field_name,
          api_id
        )
      )
      return
    end
  end
  -- fall through
  request_handle:streamInfo():dynamicMetadata():set("proxyconf.downstream_auth", "status", "failed")
  request_handle:logInfo(
    string.format(
      "Downstream authentication (%s:%s) failed for API %s, due to no config found",
      auth_type,
      auth_field_name,
      api_id
    )
  )
end

function envoy_on_response(response_handle) end

File: lib/proxyconf/downstream_auth.ex
defmodule ProxyConf.DownstreamAuth do
  require Logger
  defstruct([:api_id, :auth_type, :auth_field_name, :hashes, :jwt_provider_config])

  @downstream_auth_extension_key "x-proxyconf-downstream-auth"
  def to_config(_api_id, %{@downstream_auth_extension_key => "disabled"}), do: :disabled

  def to_config(api_id, %{
        @downstream_auth_extension_key => %{
          "auth_type" => "header",
          "config" => %{"name" => header_name, "hashes" => hashes}
        }
      }),
      do: %__MODULE__{
        api_id: api_id,
        auth_type: "header",
        auth_field_name: header_name,
        hashes: hashes
      }

  def to_config(api_id, %{
        @downstream_auth_extension_key => %{
          "auth_type" => "query",
          "config" => %{"name" => query_field_name, "hashes" => hashes}
        }
      }),
      do: %__MODULE__{
        api_id: api_id,
        auth_type: "query",
        auth_field_name: query_field_name,
        hashes: hashes
      }

  @jwt_provider_ext "extensions.filters.http.jwt_authn.v3.JwtProvider"

  def to_config(api_id, %{
        @downstream_auth_extension_key => %{
          "auth_type" => "jwt",
          "config" => jwt_provider_config
        }
      }),
      do: %__MODULE__{api_id: api_id, auth_type: "jwt", jwt_provider_config: jwt_provider_config}

  def to_config(_api_id, _spec) do
    raise(
      "API doesn't configure downstream authentication, which isn't allowed. To disable downstream authentication (not recommended) you can specify '#{@downstream_auth_extension_key}' to 'disabled'"
    )
  end

  def to_filter_metadata(api_id, spec) do
    case to_config(api_id, spec) do
      %__MODULE__{} = config ->
        %{
          "api_id" => api_id,
          "auth_type" => config.auth_type,
          "auth_field_name" => config.auth_field_name
        }

      _ ->
        %{}
    end
  end

  @external_resource "lua/vendor/md5/md5.lua"
  @lua_includes Enum.reduce(["lua/vendor/md5/md5.lua"], [], fn f, code ->
                  module_name = Path.basename(f) |> String.replace_suffix(".lua", "")
                  module_code = File.read!(f)

                  module_loader = """
                  package.loaded["#{module_name}"] = package.loaded["#{module_name}"] or (function(...) #{module_code} end)("#{module_name}") or package.loaded["#{module_name}"] or true
                  """

                  [module_loader, "\n" | code]
                end)
                |> :erlang.iolist_to_binary()

  @external_resource "downstream_auth.lua"
  @lua_prelude [
    "package.path=\"\"\n\n",
    @lua_includes,
    File.read!("lib/proxyconf/downstream_auth.lua")
  ]

  def to_envoy_http_filter(configs) do
    configs = Enum.reject(configs, fn c -> c == :disabled end)

    {jwt_configs, other_configs} =
      Enum.split_with(configs, fn config -> config.auth_type == "jwt" end)

    {providers, rules} = to_envoy_jwt_config(jwt_configs)

    if Enum.empty?(other_configs) do
      [
        %{
          "name" => "envoy.filters.http.jwt_authn",
          "typed_config" => %{
            "@type" =>
              "type.googleapis.com/envoy.extensions.filters.http.jwt_authn.v3.JwtAuthentication",
            "providers" => providers,
            "rules" => rules
          }
        }
      ]
    else
      [
        %{
          "name" => "envoy.filters.http.jwt_authn",
          "typed_config" => %{
            "@type" =>
              "type.googleapis.com/envoy.extensions.filters.http.jwt_authn.v3.JwtAuthentication",
            "providers" => providers,
            "rules" => rules
          }
        },
        %{
          "name" => "envoy.filters.http.lua",
          "typed_config" => %{
            "@type" => "type.googleapis.com/envoy.extensions.filters.http.lua.v3.Lua",
            "default_source_code" => %{
              "inline_string" => to_lua(configs)
            }
          }
        }
      ]
    end
  end

  defp to_envoy_jwt_configs(configs) do
    Enum.group_by(configs, fn config -> config.jwt_provider_config end, fn config ->
      %{
        api_id: config.api_id
      }
    end)
    |> Enum.reduce({%{}, []}, fn {provider_config, api_id_and_rules},
                                 {providers_acc, rules_acc} ->
      provider_name =
        "jwt-provider-#{:crypto.hash(:md5, :erlang.term_to_binary(provider_config)) |> Base.encode(16) |> String.slice(0, 10)}"
  {Map.put(providers_acc, provider_name, provider_config), 
    end)
  end

  defp to_lua(configs) do
    [
      @lua_prelude
      | Enum.map_join(
          configs,
          "\n",
          fn %__MODULE__{} = config ->
            hashes =
              Enum.map_join(config.hashes, ",\n  ", fn hash -> "[\"#{hash}\"] = {allow=true}" end)

            """
            Config[to_key("#{config.api_id}", "#{config.auth_type}", "#{config.auth_field_name}")] = {#{hashes}}
            """
          end
        )
    ]
    |> :erlang.iolist_to_binary()
  end
end
